components/identity-core/org.wso2.carbon.identity.core/src/test/java/org/wso2/carbon/identity/core/SAMLSSOServiceProviderManagerTest.java

```diff
@@ -0,0 +1,356 @@
+        assertEquals(serviceProviderFromStorage, sampleServiceProvider1);
+    }
+
+    @Test
+    public void testAddEmptyServiceProvider() {
+
+        assertThrows(IdentityException.class,
+                () -> samlSSOServiceProviderManager.addServiceProvider(invalidServiceProviderDO, TENANT_ID));
+    }
+
+    @Test
+    public void addServiceProviderWithDuplicateIssuer() throws Exception {
+
+        samlSSOServiceProviderManager.addServiceProvider(sampleServiceProvider1, TENANT_ID);
+        System.out.println("sampleServiceProvider iss " + sampleServiceProvider1);
+        assertFalse(samlSSOServiceProviderManager.addServiceProvider(sampleServiceProvider1, TENANT_ID));
+    }
+
+    @Test
+    public void testAddServiceProviderWithException() throws Exception {
+
+        SAMLSSOServiceProviderDO serviceProviderDO = createServiceProviderDO(ISSUER1);
+
+        try (MockedStatic<JdbcUtils> jdbcUtilsMockedStatic = mockStatic(JdbcUtils.class)) {
+            NamedJdbcTemplate namedJdbcTemplate = mock(NamedJdbcTemplate.class);
+            jdbcUtilsMockedStatic.when(JdbcUtils::getNewNamedJdbcTemplate).thenReturn(namedJdbcTemplate);
+            doThrow(new TransactionException("Transaction failed")).when(namedJdbcTemplate).withTransaction(any());
+
+            assertThrows(IdentityException.class,
+                    () -> samlSSOServiceProviderManager.addServiceProvider(serviceProviderDO, TENANT_ID));
+        }
+    }
+
+    @Test
+    public void testUpdateServiceProvider() throws Exception {
+
+        samlSSOServiceProviderManager.addServiceProvider(sampleServiceProvider1, TENANT_ID);
+        sampleServiceProvider1.setDoSingleLogout(UPDATED_DO_SINGLE_LOGOUT);
+        sampleServiceProvider1.setRequestedRecipients(UPDATED_REQUESTED_RECIPIENTS);
+
+        samlSSOServiceProviderManager.updateServiceProvider(sampleServiceProvider1, getIssuerWithQualifier(ISSUER1),
+                TENANT_ID);
+
+        SAMLSSOServiceProviderDO updatedServiceProvider =
+                samlSSOServiceProviderManager.getServiceProvider(getIssuerWithQualifier(ISSUER1), TENANT_ID);
+        assertEquals(sampleServiceProvider1, updatedServiceProvider);
+    }
+
+    @Test
+    public void testUpdateServiceProviderWithDuplicateIssuer() throws Exception {
+
+        sampleServiceProvider1 = createServiceProviderDO(ISSUER1);
+        sampleServiceProvider2 = createServiceProviderDO(ISSUER2);
+        samlSSOServiceProviderManager.addServiceProvider(sampleServiceProvider1, TENANT_ID);
+        samlSSOServiceProviderManager.addServiceProvider(sampleServiceProvider2, TENANT_ID);
+
+        sampleServiceProvider2.setIssuer(getIssuerWithQualifier(ISSUER1));
+        assertFalse(samlSSOServiceProviderManager.updateServiceProvider(sampleServiceProvider2, ISSUER2, TENANT_ID));
+    }
+
+    @Test
+    public void testUpdateEmptyServiceProvider() {
+
+        assertThrows(IdentityException.class,
```

Comment:
Don't we need to add the service provide before do the update?

Reply:
Resolved in [Introduce JDBC based persistence for SAML](https://github.com/wso2/carbon-identity-framework/pull/6241/commits/d1e67ce649f7a5d2e7e62f364ce93815f347e7b4)